### ISSUE TITLE
panic: expected newly cut file to have sequence:offset

### ISSUE THREAD

**Issue by chrismwendt** [prometheus/prometheus#10291]
**Environment**

* System information: Darwin 20.6.0 arm64

* Prometheus version:

```
prometheus, version 2.33.1 (branch: HEAD, revision: 4e08110891fd5177f9174c4179bc38d789985a13)
  build user:       root@679bbc9ae92c
  build date:       20220202-15:09:10
  go version:       go1.17.6
  platform:         linux/arm64
```

* Logs:

```
t=2022-02-11T21:59:50+0000 lvl=dbug msg="TRACE internal" host=host.docker.internal:3090 path=/.internal/configuration code=200 duration=7.476083ms
t=2022-02-11T21:59:55+0000 lvl=dbug msg="TRACE internal" host=host.docker.internal:3090 path=/.internal/configuration code=200 duration=4.621958ms
t=2022-02-11T21:59:58+0000 lvl=dbug msg="TRACE internal" host=host.docker.internal:3090 path=/.internal/configuration code=200 duration=3.790625ms
panic: expected newly cut file to have sequence:offset 2:8, got 1:8

goroutine 427 [running]:
github.com/prometheus/prometheus/tsdb.handleChunkWriteError({0x3218ec0, 0x40020e12a8})
	/app/tsdb/head_append.go:585 +0x98
github.com/prometheus/prometheus/tsdb/chunks.(*chunkWriteQueue).processJob(0x4000192100, {0x1, 0xec36, 0x17eeab75431, 0x17eeacc62c1, {0x328aa10, 0x4001d734c0}, 0x200000008, 0x2deb8e8})
	/app/tsdb/chunks/chunk_write_queue.go:106 +0x90
github.com/prometheus/prometheus/tsdb/chunks.(*chunkWriteQueue).start.func1(0x4000192100)
	/app/tsdb/chunks/chunk_write_queue.go:94 +0xc4
created by github.com/prometheus/prometheus/tsdb/chunks.(*chunkWriteQueue).start
	/app/tsdb/chunks/chunk_write_queue.go:90 +0x58
t=2022-02-11T22:00:00+0000 lvl=eror msg="command [/prometheus.sh --web.listen-address=0.0.0.0:9092] exited: exit status 2" cmd=prom-wrapper
```

**Comment by replacewithname** [comment]
Don't know if it helps:

I had the same error message yesterday on arm64 Kubernetes with version 2.33.3.

I downgraded to 2.32.1 to see wether the minor version caused the bug. The container started by reapplying the WAL and then failing with a "no space left on device" error. My PersistentVolume was exceeded.

I extended the PV and the corresponding claim, restarted the pod and the WAL was successfully reapplied, everything is back to normal now.

I'm still on 2.32.1 so I don't know if upgrading would cause the same issue. But downgrading and / or looking for storage issues might be a hint for this problem.

**Comment by hhoffstaette** [comment]
This started happening with 2.33.0 and has nothing to do with storage being faulty; 2.32.x indeed works fine.

**Comment by roidelapluie** [comment]
@codesome could this be linked to #10051 ?

**Comment by codesome** [comment]
@roidelapluie yes. We faced this issue as well and working on a patch and will require a patch release. We think it is because of failed m-map chunk replays.

@chrismwendt could you share some more logs right before the panic happened? (to confirm our hypothesis)

**Comment by codesome** [comment]
Our hypothesis that it might be because of failed m-map chunks might not be right (if m-map files were deleted during the replay, then there is still a bug and will be fixed, but in our particular case, no m-map files were deleted by the repair). We are still investigating.

**Comment by robertauer** [comment]
I experienced the same issue with Prometheus 2.33.3 four days ago. Restarted it and until now it hasn't occurred again.

**Comment by codesome** [comment]
@chrismwendt @robertauer could you maybe share your Prometheus config and CLI flags? Mainly looking at the scrape interval and other TSDB settings.

**Comment by codesome** [comment]
I was able to reproduce it https://gist.github.com/codesome/326738056f5cef48563f25d4ad4c0af3

In our case, we also saw failed m-map chunks replay, but that happened after the panic. So the above test reproduces the following hypothesis:
1. The scrape interval is >30m. Which makes a chunk to be >1h long.
2. So when there is >2h of data, but <3h, 2 chunks would be m-mapped, remaining data is in the memory. And the last m-map file is 001.
3. A restart happens due to an upgrade. So now, a new m-map file is not created until a new chunk comes for m-mapping. So the queue keeps "current file" as 001, hence when we are going to cut a new file, it will predict 002 for the new chunk.
4. But then the data crossed 3h, so a block was compacted, along with that the m-map file 001 was also deleted since all the data in that file has been compacted.
5. At this point, the queue's "current file" is still 001. But when we go to cut a new m-map file, it looks at the disk, sees no file, and creates a file 001, while the queue was expecting 002 for the new file.
6. Hence the panic.

Solution: Update the current file reference in the queue whenever the files are deleted.

---

@replay found another way this panic could happen. If there was a m-map file corruption that ends up deleting the last file (or more) on disk, the "current file" is still not updated in the queue. Hence the next file prediction fails again.

---

I will work on a fix to fix both the things.

**Comment by codesome** [comment]
Closed by https://github.com/prometheus/prometheus/pull/10316, will be fixed in v2.33.4

**Comment by pierluigilenoci** [comment]
Hi @codesome,
we have the same problem also with version v2.33.4.

```shell
ts=2022-03-11T15:05:12.501Z caller=main.go:512 level=info msg="Starting Prometheus" version="(version=2.33.4, branch=HEAD, revision=83032011a5d3e6102624fe58241a374a7201fee8)"
ts=2022-03-11T15:05:12.501Z caller=main.go:517 level=info build_context="(go=go1.17.7, user=root@d13bf69e7be8, date=20220222-16:51:28)"
ts=2022-03-11T15:05:12.501Z caller=main.go:518 level=info host_details="(Linux 5.4.0-1055-azure #57~18.04.1-Ubuntu SMP Fri Jul 16 19:40:19 UTC 2021 x86_64 prometheus-server-58bc9749d6-p2pwl (none))"
ts=2022-03-11T15:05:12.501Z caller=main.go:519 level=info fd_limits="(soft=1048576, hard=1048576)"
ts=2022-03-11T15:05:12.501Z caller=main.go:520 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2022-03-11T15:05:12.506Z caller=web.go:570 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
ts=2022-03-11T15:05:12.508Z caller=main.go:923 level=info msg="Starting TSDB ..."
ts=2022-03-11T15:05:12.508Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1645690147920 maxt=1645725600000 ulid=01FWPZEFB9JQP1E35YK3V6B4S0
ts=2022-03-11T15:05:12.508Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1645725600497 maxt=1645790400000 ulid=01FWRX8AZRP3HQAQDJSH8N8KGK
ts=2022-03-11T15:05:12.508Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1645790400191 maxt=1645855200000 ulid=01FWTV1DN0ADMJPN3CA2MXNPEC
ts=2022-03-11T15:05:12.508Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1645855200293 maxt=1645920000000 ulid=01FWWRV76G1J6HY2YSGRHBB842
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1645920000298 maxt=1645984800000 ulid=01FWYPMG1PFZBA060P34G4D7F4
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1645984800048 maxt=1646049600000 ulid=01FX0MES47AHYA7KBC7DJPT28H
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646049600295 maxt=1646114400000 ulid=01FX2J7JN4BH9NTECQ74XPX19X
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646114400497 maxt=1646179200000 ulid=01FX4G1BPT3C24VC8YXSHX55GR
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646179200261 maxt=1646244000000 ulid=01FX6DTM58WF1DHEPKKFGJ4JC8
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646244000067 maxt=1646308800000 ulid=01FX8BMFWGKWVX3AHXYHEXRQ2N
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646308800379 maxt=1646373600000 ulid=01FXA9DV2N3XNTMT9Y4ND40WJ8
ts=2022-03-11T15:05:12.509Z caller=tls_config.go:195 level=info component=web msg="TLS is disabled." http2=false
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646373600142 maxt=1646438400000 ulid=01FXC77PF9MNNEQFCVAAH33BZY
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646438400415 maxt=1646503200000 ulid=01FXE50T8XC6ZQZXJ32NZ6WFPY
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646503200400 maxt=1646568000000 ulid=01FXG2TSRE58XDZPQ70Q2T9E4Z
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646568000254 maxt=1646632800000 ulid=01FXJ0M1QH2V9EJB8SFBA6PYFV
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646632800252 maxt=1646697600000 ulid=01FXKYDV2NT662EG2MENCA07N6
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646697600170 maxt=1646762400000 ulid=01FXNW6ZYQG1F36TMVZABZNYWY
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646827200000 maxt=1646834400000 ulid=01FXQK3BHJJ379RD9TZEQZAHM1
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646834400000 maxt=1646841600000 ulid=01FXQTNSZVTGX4SVNWJS59E9H0
ts=2022-03-11T15:05:12.509Z caller=repair.go:57 level=info component=tsdb msg="Found healthy block" mint=1646762400497 maxt=1646827200000 ulid=01FXQV85HN4CVW0KHYS5CYV73K
ts=2022-03-11T15:05:13.099Z caller=head.go:493 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2022-03-11T15:05:14.006Z caller=head.go:527 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=906.482762ms
ts=2022-03-11T15:05:14.006Z caller=head.go:533 level=info component=tsdb msg="Replaying WAL, this may take a while"
ts=2022-03-11T15:07:40.460Z caller=head.go:569 level=info component=tsdb msg="WAL checkpoint loaded"
ts=2022-03-11T15:07:47.002Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=173 maxSegment=559
ts=2022-03-11T15:07:57.197Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=174 maxSegment=559
ts=2022-03-11T15:07:57.448Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=175 maxSegment=559
ts=2022-03-11T15:07:57.448Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=176 maxSegment=559
ts=2022-03-11T15:08:04.971Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=177 maxSegment=559
ts=2022-03-11T15:08:06.249Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=178 maxSegment=559
ts=2022-03-11T15:08:06.250Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=179 maxSegment=559
ts=2022-03-11T15:08:24.211Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=180 maxSegment=559

panic: expected newly cut file to have sequence:offset 304:8, got 305:8

goroutine 410 [running]:
github.com/prometheus/prometheus/tsdb.handleChunkWriteError({0x35efba0, 0xc05b80aed0})
	/app/tsdb/head_append.go:585 +0x76
github.com/prometheus/prometheus/tsdb/chunks.(*chunkWriteQueue).processJob(0xc000131680, {0x1, 0x461c2, 0x17f6facbce9, 0x17f6fd5db70, {0x366a4d8, 0xc050469c80}, 0x13000000008, 0x32224d8})
	/app/tsdb/chunks/chunk_write_queue.go:106 +0xbf
github.com/prometheus/prometheus/tsdb/chunks.(*chunkWriteQueue).start.func1()
	/app/tsdb/chunks/chunk_write_queue.go:94 +0x165
created by github.com/prometheus/prometheus/tsdb/chunks.(*chunkWriteQueue).start
	/app/tsdb/chunks/chunk_write_queue.go:90 +0x72
```

If you need any other information to help us, I am fully available.

Thanks in advance.

**Comment by replay** [comment]
Hi @pierluigilenoci, thanks for reporting the issue. Do you roughly know how long after starting the Prometheus instance the panic has occurred? Did it happen like right after starting it, or after a significant amount of time has passed? I'm trying to figure out whether this is a bug related to the initialization procedure or not.

**Comment by pierluigilenoci** [comment]
@replay practically right after booting. If you look closely at the log you can read the Prometheus boot strings:
```shell
ts=2022-03-11T15:05:12.501Z caller=main.go:512 level=info msg="Starting Prometheus" version="(version=2.33.4, branch=HEAD, revision=83032011a5d3e6102624fe58241a374a7201fee8)"
```
and after a couple of minutes, during the WAL load:
```shell
ts=2022-03-11T15:08:24.211Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=180 maxSegment=559
panic: expected newly cut file to have sequence:offset 304:8, got 305:8
```
After the reboot, it happens again but maybe on some different chunk.

**Comment by replay** [comment]
Thanks for your reply!

There is no guarantee that the panic on the second line happened right after the log statement on the first line of this snippet, that's why I had to ask about the timing:
```
ts=2022-03-11T15:08:24.211Z caller=head.go:604 level=info component=tsdb msg="WAL segment loaded" segment=180 maxSegment=559
panic: expected newly cut file to have sequence:offset 304:8, got 305:8
```

> After the reboot, it happens again but maybe on some different chunk.

This is very interesting, I have two more questions:

1) When you restart your Prometheus is there always a panic, or just sometimes?
2) Are you sure that you don't have multiple instances of Prometheus writing into the same data directory? <- that's a long shot, but it would explain the behavior you're describing...

**Comment by pierluigilenoci** [comment]
@replay I'm sure the panic happened right after the previous line because I saw it happen with my own eyes.
I was checking the live logs during `CrashLoopBack`.

The situation was created in this way:
* During the Prometheus update ([Chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus) from 15.4.0 to 15.5.3, Prometheus from v2.33.4 to v2.33.5) the updated server pod started complaining about the DB lock.
* We added the flag `storage.tsdb.no-lockfile` to resolve.
* The pods started rebooting giving that error.

Note: we installed Prometheus with these values in the chart exactly to never have two pods running at the same time. 
We save the data on a PVC and another update strategy didn't work because the pods were `fighting` to get the same disk.

```yaml
strategy:
  type: Recreate
podDisruptionBudget:
  enabled: true
  maxUnavailable: 1
```

The error was resolved by increasing the CPU/memory limit of the pod to 3Core/8GB.
Maybe it was a coincidence but this solved it.
